diem-move/parallel-executor/src/scheduler.rs

```diff
@@ -1,113 +1,469 @@
+    // may return a re-execution task back to the caller (otherwise, NoTask).
+    pub fn finish_validation(
+        &self,
+        txn_idx: TxnIndex,
+        incarnation: Incarnation,
+        aborted: bool,
+    ) -> SchedulerTask {
+        if aborted {
+            self.set_aborted_status(txn_idx, incarnation);
+
+            // Schedule strictly higher txns for validation
+            // (txn_idx needs to be re-executed first).
+            self.decrease_validation_idx(txn_idx + 1);
+
+            // txn_idx must be re-executed, and if execution_idx is lower, it will be.
+            if self.execution_idx.load(Ordering::SeqCst) > txn_idx {
+                // Optimization: execution_idx is higher than txn_idx, but decreasing it may
+                // lead to wasted work for all indices between txn_idx and execution_idx.
+                // Instead, attempt to create a new incarnation and return the corresponding
+                // re-execution task back to the caller. If incarnation fails, there is
+                // nothing to do, as another thread must have succeeded to incarnate and
+                // obtain the task for re-execution.
+                if let Some(new_incarnation) = self.incarnate(txn_idx) {
+                    return SchedulerTask::ExecutionTask((txn_idx, new_incarnation));
+                }
+            }
+        }
+
+        // Decrement active tasks to account for the finished validation.
+        assert!(self.num_active_tasks.fetch_sub(1, Ordering::SeqCst) > 0);
+        SchedulerTask::NoTask
+    }
+
+    // Reset txn_idx to end the execution earlier. The executor will stop at the smallest
+    // `stop_idx` when there are multiple concurrent invocation.
+    pub fn set_stop_idx(&self, stop_idx: TxnIndex) {
+        self.stop_idx.fetch_min(stop_idx, Ordering::Relaxed);
+    }
+
+    pub fn num_txn_to_execute(&self) -> usize {
+        self.stop_idx.load(Ordering::Relaxed)
+    }
+
+    // A lazy, check of whether the scheduler execution is completed.
+    // Updates the 'done_marker' so other threads can know by calling done() function below.
+    //
+    // 1. After the STM execution has completed:
+    // validation_idx >= num_txn, execution_idx >= num_txn, num_active_tasks == 0,
+    // and decrease_cnt does not change - so it will be successfully detected.
+    // 2. If done_marker is set, all of these must hold at the same time, implying completion.
+    // Proof: O.w. one of the indices must decrease from when it is read to be >= num_txns
+    // to when num_active_tasks is read to be 0, but decreasing thread is performing an active task,
+    // so it must first perform the next instruction in 'decrease_validation_idx' or
+    // 'decrease_execution_idx' functions, which is to increment the decrease_cnt++.
+    // Final check will then detect a change in decrease_cnt and not allow a false positive.
+    fn check_done(&self, num_txns: usize) -> bool {
```

Comment:
this check feels over-complicated, can we maintain a counter of committed txn index so we know it's done when that index == stop_idx? probably need a binary indexed tree lol

Reply:
It's actually similarly non-trivial to track precisely what's committed at any given time, and doing that all the time is not cheap (and used to add significant overhead in earlier versions). This is a lazy mechanism that only checks if everything is done, and we can trigger the checks when one of the indices goes out of bounds, so not always and doesn't add much overhead now. Unless smt unexpected changed, it should be significant for the perf right now.

Complication only arises because we need to ensure that all three conditions (on num tasks and exec and val indices) hold at the same time, and these are all separately contended variables so we have to keep and read them separately. Double-collecting decrease_cnt is what allows to be safe - and the proof fits in the comments :-)

Reply:
maintaining the validation status sounds like a very classic problem for segment tree (range sum and point/range set), probably overkilling LOL

Reply:
concurrent segment trees sound cool I admit, but let's do it later? 😄